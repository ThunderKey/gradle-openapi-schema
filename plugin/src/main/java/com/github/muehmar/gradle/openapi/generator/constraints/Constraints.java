package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

public class Constraints {
  private final Min min;
  private final Max max;
  private final DecimalMin decimalMin;
  private final DecimalMax decimalMax;
  private final Size size;
  private final Pattern pattern;
  private final Email email;

  private Constraints(
      Min min,
      Max max,
      DecimalMin decimalMin,
      DecimalMax decimalMax,
      Size size,
      Pattern pattern,
      Email email) {
    this.min = min;
    this.max = max;
    this.decimalMin = decimalMin;
    this.decimalMax = decimalMax;
    this.size = size;
    this.pattern = pattern;
    this.email = email;
  }

  public static Constraints empty() {
    return new Constraints(null, null, null, null, null, null, null);
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public Constraints withMin(Min min) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withMax(Max max) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withSize(Size size) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public void onMin(Consumer<Min> onMin) {
    Optional.ofNullable(min).ifPresent(onMin);
  }

  public void onMax(Consumer<Max> onMax) {
    Optional.ofNullable(max).ifPresent(onMax);
  }

  public void onDecimalMin(Consumer<DecimalMin> onDecimalMin) {
    Optional.ofNullable(decimalMin).ifPresent(onDecimalMin);
  }

  public void onDecimalMax(Consumer<DecimalMax> onDecimalMax) {
    Optional.ofNullable(decimalMax).ifPresent(onDecimalMax);
  }

  public void onSize(Consumer<Size> onSize) {
    Optional.ofNullable(size).ifPresent(onSize);
  }

  public void onPattern(Consumer<Pattern> onPattern) {
    Optional.ofNullable(pattern).ifPresent(onPattern);
  }

  public void onEmail(Consumer<Email> onEmail) {
    Optional.ofNullable(email).ifPresent(onEmail);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Constraints that = (Constraints) o;
    return Objects.equals(min, that.min)
        && Objects.equals(max, that.max)
        && Objects.equals(size, that.size)
        && Objects.equals(pattern, that.pattern)
        && Objects.equals(email, that.email);
  }

  @Override
  public int hashCode() {
    return Objects.hash(min, max, size, pattern, email);
  }

  @Override
  public String toString() {
    return "Constraints{"
        + "min="
        + min
        + ", max="
        + max
        + ", size="
        + size
        + ", pattern="
        + pattern
        + ", email="
        + email
        + '}';
  }
}
