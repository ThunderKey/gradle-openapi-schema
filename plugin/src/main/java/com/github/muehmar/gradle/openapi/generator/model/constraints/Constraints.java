package com.github.muehmar.gradle.openapi.generator.model.constraints;

import com.github.muehmar.gradle.openapi.util.Optionals;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
@PojoBuilder
public class Constraints {
  Optional<Min> min;
  Optional<Max> max;
  Optional<DecimalMin> decimalMin;
  Optional<DecimalMax> decimalMax;
  Optional<Size> size;
  Optional<Pattern> pattern;
  Optional<Email> email;
  Optional<PropertyCount> propertyCount;
  Optional<MultipleOf> multipleOf;

  public static Constraints empty() {
    return new Constraints(
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty());
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public static Constraints ofDecimalMinAndMax(DecimalMin decimalMin, DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMin(decimalMin).withDecimalMax(decimalMax);
  }

  public static Constraints ofPropertiesCount(PropertyCount propertyCount) {
    return ConstraintsBuilder.create().andOptionals().propertyCount(propertyCount).build();
  }

  public static Constraints ofMultipleOf(MultipleOf multipleOf) {
    return ConstraintsBuilder.create().andOptionals().multipleOf(multipleOf).build();
  }

  public Constraints withMin(Min min) {
    return new Constraints(
        Optional.ofNullable(min),
        max,
        decimalMin,
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withMax(Max max) {
    return new Constraints(
        min,
        Optional.ofNullable(max),
        decimalMin,
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withSize(Size size) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        Optional.ofNullable(size),
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        size,
        Optional.ofNullable(pattern),
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        size,
        pattern,
        Optional.ofNullable(email),
        propertyCount,
        multipleOf);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(
        min,
        max,
        Optional.ofNullable(decimalMin),
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(
        min,
        max,
        decimalMin,
        Optional.ofNullable(decimalMax),
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints and(Constraints other) {
    return new Constraints(
        Optionals.or(min, min),
        Optionals.or(max, other.max),
        Optionals.or(decimalMin, other.decimalMin),
        Optionals.or(decimalMax, other.decimalMax),
        Optionals.or(size, other.size),
        Optionals.or(pattern, other.pattern),
        Optionals.or(email, other.email),
        Optionals.or(propertyCount, other.propertyCount),
        Optionals.or(multipleOf, other.multipleOf));
  }

  public Optional<Min> getMin() {
    return min;
  }

  public Optional<Max> getMax() {
    return max;
  }

  public Optional<DecimalMin> getDecimalMin() {
    return decimalMin;
  }

  public Optional<DecimalMax> getDecimalMax() {
    return decimalMax;
  }

  public Optional<Size> getSize() {
    return size;
  }

  public Optional<Pattern> getPattern() {
    return pattern;
  }

  public Optional<Email> getEmail() {
    return email;
  }

  public <R> Optional<R> onMinFn(Function<Min, R> onMin) {
    return min.map(onMin);
  }

  public <R> Optional<R> onMaxFn(Function<Max, R> onMax) {
    return max.map(onMax);
  }

  public <R> Optional<R> onDecimalMinFn(Function<DecimalMin, R> onDecimalMin) {
    return decimalMin.map(onDecimalMin);
  }

  public <R> Optional<R> onDecimalMaxFn(Function<DecimalMax, R> onDecimalMax) {
    return decimalMax.map(onDecimalMax);
  }

  public <R> Optional<R> onSizeFn(Function<Size, R> onSize) {
    return size.map(onSize);
  }

  public <R> Optional<R> onPatternFn(Function<Pattern, R> onPattern) {
    return pattern.map(onPattern);
  }
}
