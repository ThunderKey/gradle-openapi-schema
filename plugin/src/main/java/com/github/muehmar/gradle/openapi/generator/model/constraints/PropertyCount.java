package com.github.muehmar.gradle.openapi.generator.model.constraints;

import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class PropertyCount {
  private final Optional<Integer> min;
  private final Optional<Integer> max;

  private PropertyCount(Optional<Integer> min, Optional<Integer> max) {
    this.min = min;
    this.max = max;
  }

  public static PropertyCount ofMinProperties(int min) {
    return new PropertyCount(Optional.of(min), Optional.empty());
  }

  public static PropertyCount ofMaxProperties(int max) {
    return new PropertyCount(Optional.empty(), Optional.of(max));
  }

  public static PropertyCount ofMinAndMaxProperties(int min, int max) {
    return new PropertyCount(Optional.of(min), Optional.of(max));
  }

  public Optional<Integer> getMinProperties() {
    return min;
  }

  public Optional<Integer> getMaxProperties() {
    return max;
  }
}
