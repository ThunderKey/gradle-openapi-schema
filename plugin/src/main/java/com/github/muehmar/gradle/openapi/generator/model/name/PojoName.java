package com.github.muehmar.gradle.openapi.generator.model.name;

import lombok.EqualsAndHashCode;

/** Represents the name of a pojo. */
@EqualsAndHashCode
public class PojoName {
  private final Name name;
  private final String suffix;

  PojoName(Name name, String suffix) {
    this.name = name;
    this.suffix = suffix;
  }

  public static PojoName ofName(Name name) {
    return new PojoName(name.startUpperCase(), "");
  }

  public static PojoName ofNameAndSuffix(Name name, String suffix) {
    return new PojoName(name.startUpperCase(), suffix);
  }

  public static PojoName ofNameAndSuffix(String name, String suffix) {
    return new PojoName(Name.ofString(name).startUpperCase(), suffix);
  }

  public PojoName deriveMemberSchemaName(Name pojoMemberName) {
    final Name derivedName = name.startUpperCase().append(pojoMemberName.startUpperCase());
    return new PojoName(derivedName, suffix);
  }

  public PojoName appendToName(String append) {
    return new PojoName(name.append(append), suffix);
  }

  public PojoName prependSuffix(String prependText) {
    return new PojoName(name, prependText + suffix);
  }

  public boolean equalsIgnoreCase(PojoName other) {
    return name.equalsIgnoreCase(other.name) && suffix.equals(other.suffix);
  }

  public Name getName() {
    return name;
  }

  public String getSuffix() {
    return suffix;
  }

  public String asString() {
    return name.append(suffix).asString();
  }

  @Override
  public String toString() {
    return asString();
  }
}
