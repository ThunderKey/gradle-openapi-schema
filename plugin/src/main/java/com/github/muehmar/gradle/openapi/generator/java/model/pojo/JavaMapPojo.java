package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaIdentifier;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaMapType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.pojo.MapPojo;
import com.github.muehmar.gradle.openapi.generator.model.type.MapType;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaMapPojo implements JavaPojo {
  private final JavaPojoName name;
  private final String description;
  private final JavaType valueType;
  private final Constraints constraints;

  private JavaMapPojo(
      JavaPojoName name, String description, JavaType valueType, Constraints constraints) {
    this.name = name;
    this.description = description;
    this.valueType = valueType;
    this.constraints = constraints;
  }

  public static JavaMapPojo wrap(MapPojo mapPojo, TypeMappings typeMappings) {
    return new JavaMapPojo(
        JavaPojoName.wrap(mapPojo.getName()),
        mapPojo.getDescription(),
        JavaType.wrap(mapPojo.getValueType(), typeMappings),
        mapPojo.getConstraints());
  }

  @Override
  public JavaName getSchemaName() {
    return JavaName.fromName(name.getSchemaName());
  }

  @Override
  public JavaIdentifier getClassName() {
    return name.asJavaName().asIdentifier();
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public PojoType getType() {
    return PojoType.DEFAULT;
  }

  public Constraints getConstraints() {
    return constraints;
  }

  public HashCodeGenerator.HashCodeContent getHashCodeContent() {
    return HashCodeContentBuilder.create().members(PList.single(getMember())).build();
  }

  public EqualsGenerator.EqualsContent getEqualsContent() {
    return EqualsContentBuilder.create()
        .className(getClassName())
        .members(PList.single(getMember()))
        .build();
  }

  public JavaPojoMember getMember() {
    return JavaPojoMember.of(
        Name.ofString("values"), "", getMemberType(), Necessity.REQUIRED, Nullability.NOT_NULLABLE);
  }

  private JavaType getMemberType() {
    final MapType mapType = MapType.ofKeyAndValueType(StringType.noFormat(), valueType.getType());
    return JavaMapType.wrap(mapType, TypeMappings.empty());
  }

  public JavaType getValueType() {
    return valueType;
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo,
      Function<JavaComposedPojo, T> onComposedPojo,
      Function<JavaMapPojo, T> onFreeFormPojo) {
    return onFreeFormPojo.apply(this);
  }
}
