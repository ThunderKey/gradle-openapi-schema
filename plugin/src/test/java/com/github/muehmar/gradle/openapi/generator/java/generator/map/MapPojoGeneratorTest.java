package com.github.muehmar.gradle.openapi.generator.java.generator.map;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaMapPojo;
import com.github.muehmar.gradle.openapi.generator.model.PojoName;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.constraints.PropertyCount;
import com.github.muehmar.gradle.openapi.generator.model.pojo.MapPojo;
import com.github.muehmar.gradle.openapi.generator.model.type.IntegerType;
import com.github.muehmar.gradle.openapi.generator.model.type.NoType;
import com.github.muehmar.gradle.openapi.generator.model.type.ObjectType;
import com.github.muehmar.gradle.openapi.generator.settings.JsonSupport;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import com.github.muehmar.gradle.openapi.generator.settings.ValidationApi;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

@ExtendWith(SnapshotExtension.class)
class MapPojoGeneratorTest {

  public static final MapPojo OBJECT_MAP_POJO =
      MapPojo.of(
          PojoName.ofNameAndSuffix("ObjectMap", "Dto"),
          "Object map",
          NoType.create(),
          Constraints.ofPropertiesCount(PropertyCount.ofMinAndMaxProperties(5, 10)));
  public static final MapPojo DTO_MAP_POJO =
      MapPojo.of(
          PojoName.ofNameAndSuffix("DtoMap", "Dto"),
          "DTO map",
          ObjectType.ofName(PojoName.ofNameAndSuffix("Other", "Dto")),
          Constraints.ofPropertiesCount(PropertyCount.ofMinAndMaxProperties(1, 3)));

  public static final MapPojo INTEGER_MAP_POJO =
      MapPojo.of(
          PojoName.ofNameAndSuffix("FreeForm", "Dto"),
          "Free form object",
          IntegerType.formatInteger(),
          Constraints.ofPropertiesCount(PropertyCount.ofMinAndMaxProperties(5, 10)));
  private Expect expect;

  @ParameterizedTest
  @EnumSource(ValidationApi.class)
  @SnapshotName("objectMapWithDefaultSettings")
  void generate_when_defaultSettings_then_matchSnapshot(ValidationApi validationApi) {
    final MapPojoGenerator gen = new MapPojoGenerator();

    final Writer generate =
        gen.generate(
            JavaMapPojo.wrap(OBJECT_MAP_POJO, TypeMappings.empty()),
            TestPojoSettings.defaultSettings().withValidationApi(validationApi),
            Writer.createDefault());

    expect.scenario(validationApi.getValue()).toMatchSnapshot(generate.asString());
  }

  @ParameterizedTest
  @EnumSource(ValidationApi.class)
  @SnapshotName("dtoMapWithDefaultSettings")
  void generate_when_dtoMapAndDefaultSettings_then_matchSnapshot(ValidationApi validationApi) {
    final MapPojoGenerator gen = new MapPojoGenerator();

    final Writer generate =
        gen.generate(
            JavaMapPojo.wrap(DTO_MAP_POJO, TypeMappings.empty()),
            TestPojoSettings.defaultSettings().withValidationApi(validationApi),
            Writer.createDefault());

    expect.scenario(validationApi.getValue()).toMatchSnapshot(generate.asString());
  }

  @ParameterizedTest
  @EnumSource(ValidationApi.class)
  @SnapshotName("integerMapWithDefaultSettings")
  void generate_when_integerMapAndDefaultSettings_then_matchSnapshot(ValidationApi validationApi) {
    final MapPojoGenerator gen = new MapPojoGenerator();

    final Writer generate =
        gen.generate(
            JavaMapPojo.wrap(INTEGER_MAP_POJO, TypeMappings.empty()),
            TestPojoSettings.defaultSettings().withValidationApi(validationApi),
            Writer.createDefault());

    expect.scenario(validationApi.getValue()).toMatchSnapshot(generate.asString());
  }

  @Test
  @SnapshotName("objectMapWithNoJsonSupport")
  void generate_when_noJsonSupport_then_matchSnapshot() {
    final MapPojoGenerator gen = new MapPojoGenerator();

    final Writer generate =
        gen.generate(
            JavaMapPojo.wrap(OBJECT_MAP_POJO, TypeMappings.empty()),
            TestPojoSettings.defaultSettings().withJsonSupport(JsonSupport.NONE),
            Writer.createDefault());

    expect.toMatchSnapshot(generate.asString());
  }

  @Test
  @SnapshotName("objectMapWithValidationDisabled")
  void generate_when_validationDisabled_then_matchSnapshot() {
    final MapPojoGenerator gen = new MapPojoGenerator();

    final Writer generate =
        gen.generate(
            JavaMapPojo.wrap(OBJECT_MAP_POJO, TypeMappings.empty()),
            TestPojoSettings.defaultSettings().withEnableValidation(false),
            Writer.createDefault());

    expect.toMatchSnapshot(generate.asString());
  }
}
