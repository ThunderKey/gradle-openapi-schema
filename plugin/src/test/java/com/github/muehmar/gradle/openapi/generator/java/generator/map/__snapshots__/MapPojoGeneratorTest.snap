dtoMapWithDefaultSettings[jakarta-2]=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import javax.validation.Valid;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

/**
 * DTO map
 */
public final class DtoMapDto {
  @JsonValue
  private final Map<String, OtherDto> values;

  @JsonCreator
  public DtoMapDto(
      Map<String, OtherDto> values
    ) {
    this.values = values;
  }

  @Valid
  @NotNull
  public Map<String, OtherDto> getValues() {
    return values;
  }

  public Optional<OtherDto> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @Min(1)
  @Max(3)
  @JsonIgnore
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final DtoMapDto other = (DtoMapDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]


dtoMapWithDefaultSettings[jakarta-3]=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * DTO map
 */
public final class DtoMapDto {
  @JsonValue
  private final Map<String, OtherDto> values;

  @JsonCreator
  public DtoMapDto(
      Map<String, OtherDto> values
    ) {
    this.values = values;
  }

  @Valid
  @NotNull
  public Map<String, OtherDto> getValues() {
    return values;
  }

  public Optional<OtherDto> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @Min(1)
  @Max(3)
  @JsonIgnore
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final DtoMapDto other = (DtoMapDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]


integerMapWithDefaultSettings[jakarta-2]=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

/**
 * Free form object
 */
public final class FreeFormDto {
  @JsonValue
  private final Map<String, Integer> values;

  @JsonCreator
  public FreeFormDto(
      Map<String, Integer> values
    ) {
    this.values = values;
  }

  @NotNull
  public Map<String, Integer> getValues() {
    return values;
  }

  public Optional<Integer> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @Min(5)
  @Max(10)
  @JsonIgnore
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final FreeFormDto other = (FreeFormDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]


integerMapWithDefaultSettings[jakarta-3]=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * Free form object
 */
public final class FreeFormDto {
  @JsonValue
  private final Map<String, Integer> values;

  @JsonCreator
  public FreeFormDto(
      Map<String, Integer> values
    ) {
    this.values = values;
  }

  @NotNull
  public Map<String, Integer> getValues() {
    return values;
  }

  public Optional<Integer> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @Min(5)
  @Max(10)
  @JsonIgnore
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final FreeFormDto other = (FreeFormDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]


objectMapWithDefaultSettings[jakarta-2]=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

/**
 * Object map
 */
public final class ObjectMapDto {
  @JsonValue
  private final Map<String, Object> values;

  @JsonCreator
  public ObjectMapDto(
      Map<String, Object> values
    ) {
    this.values = values;
  }

  @NotNull
  public Map<String, Object> getValues() {
    return values;
  }

  public Optional<Object> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @Min(5)
  @Max(10)
  @JsonIgnore
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final ObjectMapDto other = (ObjectMapDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]


objectMapWithDefaultSettings[jakarta-3]=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * Object map
 */
public final class ObjectMapDto {
  @JsonValue
  private final Map<String, Object> values;

  @JsonCreator
  public ObjectMapDto(
      Map<String, Object> values
    ) {
    this.values = values;
  }

  @NotNull
  public Map<String, Object> getValues() {
    return values;
  }

  public Optional<Object> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @Min(5)
  @Max(10)
  @JsonIgnore
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final ObjectMapDto other = (ObjectMapDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]


objectMapWithNoJsonSupport=[
package com.github.muehmar;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

/**
 * Object map
 */
public final class ObjectMapDto {
  private final Map<String, Object> values;

  public ObjectMapDto(
      Map<String, Object> values
    ) {
    this.values = values;
  }

  @NotNull
  public Map<String, Object> getValues() {
    return values;
  }

  public Optional<Object> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @Min(5)
  @Max(10)
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final ObjectMapDto other = (ObjectMapDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]


objectMapWithValidationDisabled=[
package com.github.muehmar;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonValue;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * Object map
 */
public final class ObjectMapDto {
  @JsonValue
  private final Map<String, Object> values;

  @JsonCreator
  public ObjectMapDto(
      Map<String, Object> values
    ) {
    this.values = values;
  }

  public Map<String, Object> getValues() {
    return values;
  }

  public Optional<Object> getProperty(String propertyName) {
    return Optional.ofNullable(values.get(propertyName));
  }

  /**
   * Returns the number of present properties of this object.
   */
  @JsonIgnore
  public int getPropertyCount() {
    return values.size();
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || this.getClass() != obj.getClass()) return false;
    final ObjectMapDto other = (ObjectMapDto) obj;
    return Objects.deepEquals(this.values, other.values);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
      values
    );
  }
}
]